beaker-vue/src/components/panels/PreviewPanel.vue

```diff
@@ -0,0 +1,379 @@
+                            ['csv', 'tsv', 'excel'].includes(mimeCategory(mime))
+                        ">
+                            <BeakerMimeBundle 
+                                v-if="(!isRaw && !contents.isLoading)"
+                                :mimeBundle="{[mime]: contentsWrapper}"
+                            />
+                            <CodeEditor 
+                                v-if="isRaw"
+                                display-mode="dark"
+                                :modelValue="contentsWrapper"
+                                ref="codeEditorRef"
+                                placeholder="Loading..."
+                            />
+                        </div>
+                    </div>
+                </div>
+            </div>
+        </template>
+    </Sidebar>
+</template>
+  
+<script lang="ts" setup>
+import { ref, defineProps, watch, computed, defineModel, reactive } from "vue";
+
+import PDFPreview from "../render/pdf/PDFPreview.vue";
+import Sidebar from "primevue/sidebar";
+import Toolbar from "primevue/toolbar";
+import Button from "primevue/button";
+import CodeEditor from "../misc/CodeEditor.vue";
+import BeakerMimeBundle from "../render/BeakerMimeBundle.vue";
+
+const sidebar = ref();
+const codeEditorRef = ref();
+const pdfPreviewRef = ref();
+
+// dynamically updates with result of async watch
+const fallbackMime = ref("");
+
+// should nice-looking renderers show raw instead - see config
+const isRaw = ref(false);
+
+const hookResize = () => {
+    const resize = event => {
+        const target = window.innerWidth - (event.x - dragWidth.value / 2);
+        width.value = Math.min(Math.max(target, minWidth), window.innerWidth * 0.9);
+    };
+    const unhookResize = () => {
+        document.querySelector('body').removeEventListener('mousemove', resize);
+        document.querySelector('body').removeEventListener('mouseup', this);
+    };
+    document.querySelector('body').addEventListener('mousemove', resize);
+    document.querySelector('body').addEventListener('mouseup', unhookResize);
+}
+
+type PreviewCategory =
+    | 'image' 
+    | 'csv' 
+    | 'tsv'
+    | 'plaintext' 
+    | 'pdf'
+    | 'excel'
+
+// text/csv should have more priority than text/?? or octet-stream for rich matching.
+const mimeCategory = (mimetype: string): PreviewCategory => {
+    if (mimetype.startsWith('image/')) {
+        return 'image'
+    }
+    if (mimetype === 'application/pdf') {
+        return 'pdf'
+    }
+    if (mimetype === 'text/csv') {
+        return 'csv'
+    }
+    if (mimetype === 'text/tsv') {
+        return 'tsv'
+    }
+    // todo: is this comprehensive?
+    if (mimetype === 'application/vnd.ms-excel' || 
+        mimetype === 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet') 
+    {
+        return 'excel'
+    }
+    // application/octet-stream is a common plaintext unknown
+    return 'plaintext'
```

Comment:
Binary/octet files are kind of extra ugly. Maybe have these default to "Binary file, contents not shown", but with the "raw" button to force it to show?

Reply:
unfortunately, most files e.g. readme, dockerfiles, everything else show as application/octet-stream! jupyter gives "null" from file panel data, and the head fetch shows this. i don't see a good way to guess at what's string encoded.